### ISSUE TITLE
build_runner takes over 2 min to create a reflectable file

### ISSUE THREAD

**Issue by efreibe** [google/reflectable.dart#188]
Recently I needed to update the reflectable file because a property in one of my classes changed.

I had to update the pubspec.yaml because I was using build_runner 1.0.0 and it is not supported anymore.

So I'm using reflectable 2.1.0 and latest version of build_runner.

I remember that, some months ago, the same file was built in a couple of seconds.
Now it last more than two minutes for the same file.

I have discovered that, the more capabilities you add to the class, the delay is greater.

```dart
class Reflector extends Reflectable {
  const Reflector()
      : super(
          newInstanceCapability,
          invokingCapability,
          reflectedTypeCapability,
          typingCapability,
          declarationsCapability,
        );
}
```

```log
[INFO] 2m 29s elapsed, 0/1 actions completed.
[INFO] 2m 30s elapsed, 0/1 actions completed.
[WARNING] No actions completed for 15.1s, waiting on:
  - reflectable:reflectable on lib/parser/ReflectableClasses.dart

[INFO] 2m 31s elapsed, 0/1 actions completed.
[INFO] 2m 32s elapsed, 0/1 actions completed.
[INFO] 2m 33s elapsed, 0/1 actions completed.
[INFO] 2m 34s elapsed, 0/1 actions completed.
[INFO] 2m 35s elapsed, 0/1 actions completed.
[INFO] Running build completed, took 2m 36s

[INFO] Caching finalized dependency graph...
[INFO] Caching finalized dependency graph completed, took 32ms

[INFO] Succeeded after 2m 36s with 1 outputs (1 actions)
```

build.yaml
```
targets:
  $default:
    builders:
      reflectable:
        generate_for:
          - lib/parser/ReflectableClasses.dart
        options:
          formatted: false
```

flutter doctor -v
```
[✓] Flutter (Channel master, v1.10.7-pre.127, on Mac OS X 10.14.6 18G95, locale en-AR)
    • Flutter version 1.10.7-pre.127 at /Users/ernestofreiberg/flutter
    • Framework revision 75c07ef1a6 (3 hours ago), 2019-10-02 15:36:07 -0700
    • Engine revision aa8fcbb652
    • Dart version 2.6.0 (build 2.6.0-dev.0.0 98b286de21)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/ernestofreiberg/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.0, Build version 11A420a
    • CocoaPods version 1.7.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 37.1.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] VS Code (version 1.38.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.5.1

[✓] Connected device (4 available)
    • Android SDK built for x86 64 • emulator-5554   • android-x64    • Android 9 (API 28) (emulator)
    • macOS                        • macOS           • darwin-x64     • Mac OS X 10.14.6 18G95
    • Chrome                       • chrome          • web-javascript • Google Chrome 77.0.3865.90
    • Headless Server              • headless-server • web-javascript • Flutter Tools

• No issues found!
```

**Comment by eernstg** [comment]
> the same file was built in a couple of seconds.
> Now it last more than two minutes for the same file.

First thoughts on this:

The code generation step can be incremental (in the sense that `build_runner` may use information about the input to earlier code generation jobs to omit certain files from the input to the reflectable builder), so the build step may be much quicker in some situations. This could very well explain the large difference you mention.

However, the fact that reflectable is able to quantify over all subtypes of a given type (and other, similar things) implies that it must be a closed world analysis. For a given entry point _P_ (that is, a given "main" file), it must perform the analysis on every library which is reachable by any chain of import clauses from _P_. So a change in any part of a program will cause a new closed world analysis to be performed.

In some cases you may happen to know that the given changes will not invalidate the analysis performed by reflectable (e.g., because you only changed things that will never be subject to reflective access). In that situation could be useful for you to disable the build step (or if invoked manually, to simply postpone the build step until something indicates that a new run is needed badly enough).

**Comment by efreibe** [comment]
I see. The only thing that appears to be a little weird is that the same file with same dependencies was built in few seconds while now, with the packages updated, lasts more than 2 minutes.

But I think that maybe any dependency of my dependencies is creating some sort of delay...

There is a way that I could trace the build process?

**Comment by eernstg** [comment]
Reflectable recently had to be rewritten extensively in order to perform certain operations asynchronously (because the analyzer APIs changed), but that should actually allow for doing synchronous work during the time where some asynchronous operation is waiting for some external event, that is, it should be faster if anything.

> the more capabilities you add to the class, the delay is greater.

That is to be expected: More capabilities means more work, and more generated code.

But it would be interesting to have some information about which specific steps are slow.

> There is a way that I could trace the build process?

Have you tried to pass `--verbose`, as in `pub run build_runner  build --verbose something`?

**Comment by efreibe** [comment]
Unluckily the command 'flutter packages pub run build_runner build --verbose' does not change the output that is sent to console.

I thought if there was some way to debug that code, but I do not understand so well how build_runner works. Trying to add some output to the code fails because apparently the code is compiled. I'm sorry for that.

**Comment by eernstg** [comment]
@efreibe wrote:
> '.. --verbose' does not change the output

Ah, that's unfortunate. It was a source of additional information a while ago, but it might have been changed, or it might work differently under flutter than it does in a plain command line context. Thanks for trying it out, though!

Looking at the output you [included](https://github.com/dart-lang/reflectable/issues/188#issue-501803924), it seems likely that you'd get something like the following message every 10-20 seconds?:

```
[WARNING] No actions completed for 15.1s, waiting on:
  - reflectable:reflectable on lib/parser/ReflectableClasses.dart
```

Also, the following output implies that there is just one generated file, which means that the reflectable builder is executed once (if it isn't skipped entirely because everything is up to date already):

```
[INFO] Succeeded after 2m 36s with 1 outputs (1 actions)
```

How much code does the generated file contain?

**Comment by Brooooooklyn** [comment]
same issue here

**Comment by efreibe** [comment]
> Looking at the output you included, it seems likely that you'd get something like the following message every 10-20 seconds?:

Yes, apparently the builder is outputting every 10 seconds or so that message.

> How much code does the generated file contain?

As you said is one file with ~3000 lines. But this has changed as little as 20 lines in the last 6 months. Reflected file is ~28000 lines.

**Comment by eernstg** [comment]
The [test suite](https://github.com/dart-lang/test_reflectable) allows for generating 25229 lines of code in  about 17 seconds:

```
Succeeded after 17.1s with 348 outputs (214 actions)
```

so we definitely don't have the factor of 80 slowdown that you mention (2 seconds --> 2m 36s or so).

The rewrite that introduced a lot of asynchronous computation should not have caused any algorithms that I can think of to be replaced by different algorithms with a much worse complexity, but it would still be useful to know more about the situation where the very slow processing takes place.

@efreibe, @Brooooooklyn, could you make it possible for others to run the generation step and see what's going on? (That is, e.g., is your code open source?)

**Comment by efreibe** [comment]
> @efreibe, @Brooooooklyn, could you make it possible for others to run the generation step and see what's going on? (That is, e.g., is your code open source?)

It's not open source but I'm attaching the file is giving me problems, I've stripped out the libraries and made mock classes to replace the original ones. The only imports that have been left are meta and reflectable.

> [INFO] Build:Succeeded after 1m 59s with 1 outputs (1 actions)

Still I'm getting 2mins build time. 

[https://gist.github.com/efreibe/8e7f7d1e1b8075dd69546f18ae27de10](https://gist.github.com/efreibe/8e7f7d1e1b8075dd69546f18ae27de10)

**Comment by eernstg** [comment]
I get a quite different timing:

```
dart/scratch/201910/reflectable188> pub run build_runner build bin
[INFO] Generating build script completed, took 445ms
[WARNING] Deleted previous snapshot due to missing asset graph.
[INFO] Creating build script snapshot... completed, took 21.2s
[INFO] Building new asset graph completed, took 986ms
[INFO] Checking for unexpected pre-existing outputs. completed, took 4ms
[INFO] Generating SDK summary completed, took 6.0s
[INFO] reflectable:reflectable on bin/main.dart:Ignoring entry point reflectable188|bin/main.dart that does not include the library 'package:reflectable/reflect[INFO] Running build completed, took 6.3s
[INFO] Caching finalized dependency graph completed, took 49ms
[INFO] Succeeded after 6.3s with 1 outputs (1 actions)
```

and when the code generation step does not need to be performed we get this:

```
dart/scratch/201910/reflectable188> pub run build_runner build bin
[INFO] Generating build script completed, took 450ms
[INFO] Reading cached asset graph completed, took 82ms
[INFO] Checking for updates since last build completed, took 905ms
[INFO] Running build completed, took 15ms
[INFO] Caching finalized dependency graph completed, took 63ms
[INFO] Succeeded after 96ms with 0 outputs (0 actions)
```

This desktop is not particularly fancy (ok, there are a bunch of CPUs, but each is 2.90 GHz and I don't think this task uses more than one of them). So I don't think it is likely that your hardware is 20 times slower. Do you have any idea about why we'd get results that are so different?

**Comment by efreibe** [comment]
Sometime ago I've tried to install an old version of reflectable which uses an older version of build_runner and it worked like a charm. 
But now I'm restricted to new versions because flutter doesn't let me install those versions anymore.

**Comment by efreibe** [comment]
I think this is something we do not have control over...

Seems like the new analyzer is doing much more work to support
other operations, which makes it take longer to run.

**Comment by JeromePuttemans** [comment]
In my case here is the result : 
![image](https://user-images.githubusercontent.com/33145301/70078109-68125900-1602-11ea-94a0-831e7ca48483.png)

in the past (dart 2.1.0) this took more or less 2minutes, now 2 hours.

To give you an idea about my project, before dart 2.0, I created a graphical editor where I was able to update any property of an item like this : 
![image](https://user-images.githubusercontent.com/33145301/70078859-d3105f80-1603-11ea-9bfd-da39b8c355c6.png)

So I juste have to pass the instance of any item to the "object inspector" and the values of the members are displayed and updatable. As I have a lot of graphical objects, I need a lot of Reflector annotations.

May be I'll have to find a better methodology. Any idea is welcome...

**Comment by eernstg** [comment]
2h 7m, that's terrible!

We really need improvements here. It is not a normal situation that the analyzer gets a x80 slowdown, and it is also not a problem that I've heard about in all those other situations where the analyzer is used (a general x80 slowdown would cause a loud outcry from so many sources we'd all know about it immediately).

But reflectable uses the analyzer in a particular way, so maybe the new performance behavior of the analyzer is 80 times worse for reflectable, and nowhere near as bad for most other usages.

The big difference between the old and the new analyzer (that is, anything above analyzer version 0.35.0) is that the new analyzer requires asynchronous execution in order to perform certain operations. For reflectable, this occurs in particular when the analyzer is used to obtain information  to generate code for constant expressions (like the ones used in default values for parameters, or when dealing with metadata, and in a few other situations).

This means that a large number of methods needed to change from regular methods to async methods, and it means that there are a lot of context switches that we didn't have previously.

There is a lot of ongoing work (in the vm team) on improving the performance of async operations, so that might help.

**Comment by eernstg** [comment]
Another idea came up (thanks, @sigurdm!): If the huge slowdown in code generation is caused by reading source code repeatedly (because we must now use `getResolvedLibraryByElement` in order to get access to any syntactic entities, and that might cause a source file to be loaded from disk because the abstract syntax trees have already been discarded), it should help to simply fetch all resolved libraries up front and look them up in a map rather than calling `getResolvedLibraryByElement` when needed.

**Comment by HerrNiklasRaab** [comment]
Don't know if it is releated but we have also very slow build times:

First build: 11437 Lines, 336025 Characters => About 1 minute
Incremental build (with watcher): 5 - 15s 

Especially the long incremental build time is very painful to us, because you always can start test and run program, after all files are generted.

This is just for our app project. We have three other packages, in which files need to be generated. T

**Comment by HerrNiklasRaab** [comment]
To give an impression, of how slow our incremental reflectable builds are: https://www.youtube.com/watch?v=OiUBVKffTxg&feature=youtu.be

**Comment by HerrNiklasRaab** [comment]
@eernstg Sorry for bothering, but is there something we could do wrong? Our incremental builds are at about 20s sometime (mostly about 15s). We could share the project if it would help to address this.

**Comment by softmarshmallow** [comment]
any updates on this performance issue?

**Comment by eernstg** [comment]
Unfortunately no. The generation time went up when we had to change reflectable because the analyzer was changed such that a number of operations were made asynchronous (and that causes some pervasive changes because the whole call stack must then be asynchronous as well). I made a number of experiments, but nothing has come up which indicates that reflectable itself is wasting time, and the basic algorithms used by reflectable have not changed.

**Comment by k-paxian** [comment]
I think I could shed some light on this issue. Recently I've experienced build times near 2 mins

![image](https://user-images.githubusercontent.com/18071010/94998092-b3ddb800-05af-11eb-9143-e72b1b0c5afb.png)

After I did a migration of unit test files from using `part of` to plain `imports`

![image](https://user-images.githubusercontent.com/18071010/94998159-06b76f80-05b0-11eb-8a63-938994a11426.png)

**Comment by eernstg** [comment]
Thanks for the information!

Could it be the case that the 1m44s run was generating code for many tests, each of which were part of a rather large program, and with imports the code generation took place for a set of smaller programs?

So `test1` might need an import of `lib1.dart`, and `test2` might need an import of `lib2a.dart` and `lib2b.dart`; with parts, every test will have its code generated in a context where `lib1.dart`, `lib2a.dart`, `lib2b.dart` (and presumably a lot more libraries needed by `test3`, `test4`, and so on) are present; with plain imports, `test1` would have its code generated in a program that only imports whatever `test1` needs.

In that case the extra work (many reflectors in one program, each of them needs to be considered relative to the entire program) is inherent to the program, and the time spent wouldn't be caused by the introduction of a lot of asynchrony (which is the suspected culprit in this issue).

**Comment by k-paxian** [comment]
With previous/slow setup, I had one monolithic _test.dart file and a bunch of small dart files with grouped tests connected to it via "part of".

That monolithic _test.dart is the only entry point for reflectable builder.

So now the change is to replace "part of" connection with simple import statements to those small dart files.

That's basically it. My blind guess is that might be caused by analyzer code crawling through the files.

**Comment by eernstg** [comment]
Yes, that's the situation I expected. I think the substantially larger amount of work may explain the slowdown. Thanks again!

**Comment by eernstg** [comment]
I'll close this issue as obsolete, because of the many changes that have occurred in the treatment of asynchronous computations and in the analyzer. Please create new ones as needed, if you encounter similar problems.
